File: src/movieclaw_api/api/routes/subscriptions.py
from __future__ import annotations

from fastapi import APIRouter, Depends, Query
from sqlalchemy.ext.asyncio import AsyncSession

from movieclaw_api.exceptions import BadRequestException
from movieclaw_api.schemas.response import ApiResponse, ok
from movieclaw_api.schemas.subscription import (
    ActivityView,
    DispatchPreviewView,
    MediaBrief,
    PreparePayload,
    PrepareView,
    ResolveCandidateView,
    SeasonOverview,
    SubscriptionCreatePayload,
    SubscriptionDetailView,
    SubscriptionPausePayload,
    SubscriptionUpdatePayload,
    SubscriptionView,
)
from movieclaw_api.services.media_discover import get_tmdb_client
from movieclaw_api.services.media_library import MediaLibraryService
from movieclaw_api.services.subscription import SubscriptionService
from movieclaw_db.engine import get_session
from movieclaw_media.library import ResolveStatus
from movieclaw_media.models import MediaKind

router = APIRouter(prefix="/subscriptions", tags=["subscriptions"])


def _service(session: AsyncSession) -> SubscriptionService:
    library = MediaLibraryService(session, get_tmdb_client())
    return SubscriptionService(session, library)


@router.post(
    "/prepare",
    response_model=ApiResponse[PrepareView],
    summary="订阅预检：建档条目并返回季集结构（弹层数据源）",
)
async def prepare_subscription(
    payload: PreparePayload,
    session: AsyncSession = Depends(get_session),
) -> ApiResponse[PrepareView]:
    """幂等预检。TMDB 入口直接建档；豆瓣入口先收敛（命中→ready，
    歧义→candidates 让用户确认后以 tmdb_id 重新 prepare，未收录→not_found）。"""
    service = _service(session)

    if payload.source == "douban":
        if not payload.title:
            raise BadRequestException("豆瓣入口预检必须携带标题")
        library = MediaLibraryService(session, get_tmdb_client())
        resolution, item = await library.resolve_douban(
            payload.kind, payload.title, year=payload.year, douban_id=payload.douban_id
        )
        if resolution.status is ResolveStatus.NOT_FOUND:
            return ok(
                PrepareView(status="not_found"),
                message="TMDB 未收录该条目，暂无法订阅",
            )
        if resolution.status is ResolveStatus.AMBIGUOUS:
            return ok(
                PrepareView(
                    status="ambiguous",
                    candidates=[ResolveCandidateView.from_model(c) for c in resolution.candidates],
                ),
                message="找到多个可能的条目，请确认是哪一部",
            )
        assert item is not None
        tmdb_id = item.tmdb_id
    else:
        if payload.tmdb_id is None:
            raise BadRequestException("TMDB 入口预检必须携带 tmdb_id")
        tmdb_id = payload.tmdb_id

    item, seasons, existing = await service.prepare(
        payload.kind, tmdb_id, douban_id=payload.douban_id
    )
    # 库存概览（媒体库 L3 联通）：季选择器每行显示"库里已有 x 集"
    from movieclaw_db.repositories.library_file_repo import LibraryFileRepository

    assert item.id is not None
    owned = await LibraryFileRepository(session).owned_units(item.id)
    return ok(
        PrepareView(
            status="ready",
            media=MediaBrief.from_model(item),
            seasons=[SeasonOverview.from_row(s, owned_units=owned) for s in seasons],
            existing_subscription_id=existing.id if existing else None,
            movie_owned=payload.kind == MediaKind.MOVIE and (0, 0) in owned,
        )
    )


@router.post(
    "",
    response_model=ApiResponse[SubscriptionDetailView],
    summary="创建订阅（生成初始工单；同条目重复订阅幂等返回已有）",
)
async def create_subscription(
    payload: SubscriptionCreatePayload,
    session: AsyncSession = Depends(get_session),
) -> ApiResponse[SubscriptionDetailView]:
    """创建订阅。"立即踢一次缺口搜索"由 service 层统一触发，路由不用管。"""
    service = _service(session)
    subscription = await service.create(
        payload.kind,
        payload.tmdb_id,
        selected_seasons=payload.selected_seasons,
        follow_future=payload.follow_future,
        rule_set_id=payload.rule_set_id,
        library_id=payload.library_id,
        douban_id=payload.douban_id,
    )
    assert subscription.id is not None
    sub, item, wanted = await service.detail(subscription.id)
    return ok(
        SubscriptionDetailView.from_detail(sub, item, wanted),
        message="已加入订阅，正在搜索资源",
    )


@router.get(
    "/dispatch-preview",
    response_model=ApiResponse[DispatchPreviewView],
    summary="投递路由预检：按类型与目标库预演下载会落到哪、能否自动入库",
)
async def dispatch_preview(
    kind: str = Query(description="movie / tv"),
    library_id: int | None = Query(default=None, description="目标库；缺省该类型默认库"),
    session: AsyncSession = Depends(get_session),
) -> ApiResponse[DispatchPreviewView]:
    """订阅弹窗选库时调用：与真实投递同源的三级兜底 + 映射守门判定，
    配置有问题（映射不覆盖/无下载器/无库根）在订阅那一刻就亮出来。"""
    from movieclaw_api.services.download_dispatch import preview_dispatch_route

    preview = await preview_dispatch_route(session, kind=kind, library_id=library_id)
    return ok(DispatchPreviewView(**preview))


@router.get(
    "",
    response_model=ApiResponse[list[SubscriptionView]],
    summary="订阅列表（含工单进度）",
)
async def list_subscriptions(
    kind: str | None = Query(default=None, description="movie / tv，缺省全部"),
    session: AsyncSession = Depends(get_session),
) -> ApiResponse[list[SubscriptionView]]:
    service = _service(session)
    rows = await service.list_with_progress(kind=kind)
    return ok([SubscriptionView.from_model(s, m, c) for s, m, c in rows])


@router.get(
    "/{subscription_id}",
    response_model=ApiResponse[SubscriptionDetailView],
    summary="订阅详情（含工单明细）",
)
async def get_subscription(
    subscription_id: int,
    session: AsyncSession = Depends(get_session),
) -> ApiResponse[SubscriptionDetailView]:
    service = _service(session)
    sub, item, wanted = await service.detail(subscription_id)
    return ok(SubscriptionDetailView.from_detail(sub, item, wanted))


@router.get(
    "/{subscription_id}/activities",
    response_model=ApiResponse[list[ActivityView]],
    summary="订阅活动时间线（系统对该订阅做过的每个动作，时间倒序）",
)
async def list_subscription_activities(
    subscription_id: int,
    limit: int = Query(default=100, ge=1, le=500),
    session: AsyncSession = Depends(get_session),
) -> ApiResponse[list[ActivityView]]:
    service = _service(session)
    rows = await service.activities(subscription_id, limit=limit)
    return ok([ActivityView.from_model(r) for r in rows])


@router.patch(
    "/{subscription_id}",
    response_model=ApiResponse[SubscriptionDetailView],
    summary="修改订阅（季选择/追新/规则组，diff 重算工单）",
)
async def update_subscription(
    subscription_id: int,
    payload: SubscriptionUpdatePayload,
    session: AsyncSession = Depends(get_session),
) -> ApiResponse[SubscriptionDetailView]:
    service = _service(session)
    await service.update(
        subscription_id,
        selected_seasons=payload.selected_seasons,
        follow_future=payload.follow_future,
        rule_set_id=payload.rule_set_id,
        library_id=payload.library_id,
    )
    sub, item, wanted = await service.detail(subscription_id)
    return ok(SubscriptionDetailView.from_detail(sub, item, wanted), message="订阅已调整")


@router.patch(
    "/{subscription_id}/pause",
    response_model=ApiResponse[SubscriptionDetailView],
    summary="暂停 / 恢复订阅",
)
async def pause_subscription(
    subscription_id: int,
    payload: SubscriptionPausePayload,
    session: AsyncSession = Depends(get_session),
) -> ApiResponse[SubscriptionDetailView]:
    service = _service(session)
    await service.set_paused(subscription_id, payload.paused)
    sub, item, wanted = await service.detail(subscription_id)
    message = "已暂停，匹配与搜索将跳过该订阅" if payload.paused else "已恢复追踪"
    return ok(SubscriptionDetailView.from_detail(sub, item, wanted), message=message)


@router.delete(
    "/{subscription_id}",
    response_model=ApiResponse[dict],
    summary="删除订阅（不影响已下载内容）",
)
async def delete_subscription(
    subscription_id: int,
    session: AsyncSession = Depends(get_session),
) -> ApiResponse[dict]:
    service = _service(session)
    await service.delete(subscription_id)
    return ok({}, message="已取消订阅")

File: apps/web/lib/api/subscriptions.ts
import { request } from "@/lib/http";
import type { MediaType } from "@/lib/media-types";

/** 后端统一响应信封（见 movieclaw_api.schemas.response.ApiResponse） */
interface ApiEnvelope<T> {
  success: boolean;
  code: string;
  message: string;
  data: T;
}

async function unwrap<T>(promise: Promise<ApiEnvelope<T>>): Promise<T> {
  return (await promise).data;
}

/** 订阅状态（见 movieclaw_db SubscriptionStatus）：completed 是派生值 */
export type SubscriptionStatus = "active" | "paused" | "completed";

/** 工单状态机（见 movieclaw_db WantedStatus） */
export type WantedStatus = "wanted" | "grabbed" | "downloaded" | "imported";

/** 条目摘要（见 schemas.subscription.MediaBrief） */
export interface SubscriptionMedia {
  media_item_id: number;
  kind: MediaType;
  tmdb_id: number;
  douban_id: string | null;
  title: string;
  original_title: string;
  year: number | null;
  poster_url: string | null;
  status: string | null;
}

/** 弹层季选择器的一行 */
export interface SeasonOverview {
  season_number: number;
  name: string;
  air_date: string | null;
  episode_count: number | null;
  /** 已播集数（air_date<=今天） */
  aired_count: number;
  /** 媒体库已有的集数（库存 H） */
  owned_count: number;
}

/** 豆瓣收敛歧义时的确认候选 */
export interface ResolveCandidate {
  tmdb_id: number;
  title: string;
  original_title: string;
  year: number | null;
  poster_url: string | null;
}

/** 订阅预检结果：ready 可直接渲染弹层 / ambiguous 先选候选 / not_found 无法订阅 */
export interface PrepareResult {
  status: "ready" | "ambiguous" | "not_found";
  media: SubscriptionMedia | null;
  seasons: SeasonOverview[];
  existing_subscription_id: number | null;
  /** 电影：媒体库里已有本片（弹层提示，不拦订阅） */
  movie_owned: boolean;
  candidates: ResolveCandidate[];
}

export interface SubscriptionProgress {
  total: number;
  /** 缺口：还没搞到的单元数 */
  wanted: number;
  grabbed: number;
  downloaded: number;
  /** 已整理入库（终态） */
  imported: number;
}

export interface Subscription {
  id: number;
  media: SubscriptionMedia;
  status: SubscriptionStatus;
  selected_seasons: number[];
  follow_future: boolean;
  rule_set_id: number;
  /** 入库目标库；null = 该类型的默认库 */
  library_id: number | null;
  progress: SubscriptionProgress;
  created_at: string;
  updated_at: string;
}

export interface WantedItem {
  id: number;
  season_number: number;
  episode_number: number;
  status: WantedStatus;
  air_date: string | null;
  priority: number;
  next_search_at: string | null;
  search_attempts: number;
  last_search_at: string | null;
  grabbed_at: string | null;
  downloaded_at: string | null;
  imported_at: string | null;
}

export interface SubscriptionDetail extends Subscription {
  wanted: WantedItem[];
}

export interface RuleSet {
  id: number;
  name: string;
  is_default: boolean;
  spec: Record<string, unknown>;
}

export interface PreparePayload {
  source: "tmdb" | "douban";
  kind: MediaType;
  tmdb_id?: number;
  /** 豆瓣入口：豆瓣标题 */
  title?: string;
  year?: number;
  douban_id?: string;
}

export interface CreateSubscriptionPayload {
  kind: MediaType;
  tmdb_id: number;
  selected_seasons?: number[];
  follow_future?: boolean;
  rule_set_id?: number | null;
  /** 入库目标库；缺省用该类型的默认库 */
  library_id?: number | null;
  douban_id?: string | null;
}

/** 订阅预检：建档条目并返回季集结构（打开订阅弹层时调用，幂等）。 */
export function prepareSubscription(payload: PreparePayload): Promise<PrepareResult> {
  return unwrap(
    request<ApiEnvelope<PrepareResult>>("/subscriptions/prepare", {
      method: "POST",
      body: JSON.stringify(payload),
    }),
  );
}

/** 创建订阅（同条目重复订阅幂等返回已有）。 */
export function createSubscription(
  payload: CreateSubscriptionPayload,
): Promise<SubscriptionDetail> {
  return unwrap(
    request<ApiEnvelope<SubscriptionDetail>>("/subscriptions", {
      method: "POST",
      body: JSON.stringify(payload),
    }),
  );
}

/** 订阅列表（含工单进度）。kind 缺省返回全部。 */
export function listSubscriptions(
  kind?: MediaType,
  init?: RequestInit,
): Promise<Subscription[]> {
  const query = kind ? `?kind=${kind}` : "";
  return unwrap(request<ApiEnvelope<Subscription[]>>(`/subscriptions${query}`, init));
}

/** 订阅详情（含工单明细）。 */
export function getSubscription(id: number): Promise<SubscriptionDetail> {
  return unwrap(request<ApiEnvelope<SubscriptionDetail>>(`/subscriptions/${id}`));
}

/** 修改订阅（季选择/追新/规则组，后端 diff 重算工单）。 */
export function updateSubscription(
  id: number,
  payload: {
    selected_seasons?: number[];
    follow_future?: boolean;
    rule_set_id?: number;
    /** 换入库目标库；缺省不变 */
    library_id?: number;
  },
): Promise<SubscriptionDetail> {
  return unwrap(
    request<ApiEnvelope<SubscriptionDetail>>(`/subscriptions/${id}`, {
      method: "PATCH",
      body: JSON.stringify(payload),
    }),
  );
}

/** 暂停 / 恢复订阅。 */
export function pauseSubscription(id: number, paused: boolean): Promise<SubscriptionDetail> {
  return unwrap(
    request<ApiEnvelope<SubscriptionDetail>>(`/subscriptions/${id}/pause`, {
      method: "PATCH",
      body: JSON.stringify({ paused }),
    }),
  );
}

/** 删除订阅（不影响已下载内容）。 */
export function deleteSubscription(id: number): Promise<Record<string, never>> {
  return unwrap(
    request<ApiEnvelope<Record<string, never>>>(`/subscriptions/${id}`, {
      method: "DELETE",
    }),
  );
}

/** 订阅活动记录：message 是完整中文句子，时间线直接展示。 */
export interface SubscriptionActivity {
  id: number;
  type:
    | "created"
    | "adjusted"
    | "paused"
    | "resumed"
    | "completed"
    | "reopened"
    | "searched"
    | "match_accepted"
    | "match_rejected"
    | "grabbed"
    | "dispatch_failed"
    | "wanted_added";
  message: string;
  payload: Record<string, unknown>;
  created_at: string;
}

/** 订阅活动时间线（系统对该订阅做过的每个动作，时间倒序）。 */
export function listSubscriptionActivities(
  id: number,
  limit = 100,
): Promise<SubscriptionActivity[]> {
  return unwrap(
    request<ApiEnvelope<SubscriptionActivity[]>>(
      `/subscriptions/${id}/activities?limit=${limit}`,
    ),
  );
}

/** 规则组列表（首次访问后端自动创建默认组）。 */
export function listRuleSets(init?: RequestInit): Promise<RuleSet[]> {
  return unwrap(request<ApiEnvelope<RuleSet[]>>("/rule-sets", init));
}
